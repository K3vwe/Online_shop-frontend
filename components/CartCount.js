import React from "react";
import PropTypes from "prop-types";
import { TransitionGroup, CSSTransition } from "react-transition-group";
import Styled from "styled-components";

const AnimationStyle = Styled.span`
  position: relative;
  .count {
    display: block;
    position: relative;
    transition: all 0.4s;
  }
  .count-enter {
    transform: rotateX(0);
  }
  .count-enter-active{
    transform: rotateX(0.5turn);
  }
  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }
  .count-exit-active {
    transform: rotateX(0.5turn);
  }
`;

const Dot = Styled.div`
  color: white;
  background: ${(props) => props.theme.primary};
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;
  font-weight: 100;
  font-feature-settings: 'tnum';
  font-variant-numeric: tabular-nums;
`;

const CartCount = ({ count }) => (
  <AnimationStyle>
    <TransitionGroup>
      <CSSTransition
        unmountOnExit
        classNames="count"
        className="count"
        key={count}
        timeout={{ enter: 400, exit: 400 }}
      >
        <Dot>{count}</Dot>
      </CSSTransition>
    </TransitionGroup>
  </AnimationStyle>
);

export default CartCount;
