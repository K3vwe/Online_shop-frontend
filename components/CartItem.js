import React from "react";
import FormatMoney from "../lib/formatMoney";
import styled from "styled-components";
import RemoveFromCart from "./RemoveFromCart";

const StyledCartItem = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${(props) => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => {
  if (!cartItem.item)
    return (
      <StyledCartItem>
        <p>This item has been removed</p>
        <RemoveFromCart id={cartItem.id} />
      </StyledCartItem>
    );
  return (
    <StyledCartItem>
      <img width="100px" src={cartItem.item.image} alt={cartItem.item.title} />
      <div className="cart-items-details">
        <h3>{cartItem.item.title}</h3>
        <p>
          {FormatMoney(cartItem.item.price * cartItem.quantity)}
          {" - "}
          <em>
            {cartItem.quantity}
            &times;
            {FormatMoney(cartItem.item.price)}
          </em>
        </p>
      </div>
      <RemoveFromCart id={cartItem.id} />
    </StyledCartItem>
  );
};

export default CartItem;
