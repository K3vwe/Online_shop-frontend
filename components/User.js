import React from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import PropTypes from "prop-types";

export const CURRENT_USER_QUERY = gql`
  query {
    me {
      id
      email
      name
      permission
      orders {
        id
      }
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
          description
        }
      }
    }
  }
`;

const User = (props) => {
  const { data, loading } = useQuery(CURRENT_USER_QUERY);
  if (loading) return null;
  return <div {...props}>{props.children(data)}</div>;
};

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
