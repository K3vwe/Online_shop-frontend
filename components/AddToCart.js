import React from "react";
import { useMutation } from "@apollo/react-hooks";
import gql from "graphql-tag";
import { CURRENT_USER_QUERY } from "./User";

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(id: $id) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ id }) => {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });
  return (
    <button disabled={loading} onClick={addToCart}>
      Add{loading && "ing"} to Cart ðŸ›’
    </button>
  );
};

export default AddToCart;
