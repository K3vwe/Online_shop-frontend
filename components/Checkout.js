import React from "react";
import { PaystackConsumer } from "react-paystack";
import { useMutation } from "@apollo/react-hooks";
import Router from "next/router";
import NProgress from "nprogress";
import gql from "graphql-tag";
import calcTotalPrice from "../lib/calcTotalPrice";
import SickButton from "./styles/SickButton";
import User, { CURRENT_USER_QUERY } from "./User";

export const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    createOrder(token: $token) {
      id
      charge
      total
      items {
        id
        title
      }
    }
  }
`;

const totalItems = (cart) => {
  return cart.reduce((tally, cartItem) => tally + cartItem.quantity, 0);
};

const Checkout = ({ children }) => {
  const [createOrder] = useMutation(CREATE_ORDER_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  const onToken = async (res) => {
    NProgress.start();
    const order = await createOrder({
      variables: { token: res.reference },
    }).catch((error) => {
      console.log(error.message);
    });
    Router.push({
      pathname: "/order",
      query: { id: order.data.createOrder.id },
    });
  };

  return (
    <User>
      {({ me }) => {
        const componentProps = {
          reference: new Date().getTime(),
          email: me.email,
          amount: calcTotalPrice(me.cart) * 100,
          label: "Online-Shop",
          publicKey: "pk_test_e822234b0318189bdf4d1152324516f7f5b43327",
          metadata: {
            custom_field: [
              {
                display_name: "Description",
                variable_name: "description",
                value: `Order of ${totalItems(me.cart)} items!`,
              },
            ],
          },
          onSuccess: (res) => {
            console.log("====>", onToken(res));
          },
        };

        return (
          <PaystackConsumer {...componentProps}>
            {({ initializePayment }) => {
              return (
                <SickButton
                  onClick={() => {
                    initializePayment();
                  }}
                >
                  Checkout
                </SickButton>
              );
            }}
          </PaystackConsumer>
        );
      }}
    </User>
  );
};

export default Checkout;
