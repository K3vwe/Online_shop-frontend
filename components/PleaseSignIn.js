import { useQuery } from "@apollo/react-hooks";
import { CURRENT_USER_QUERY } from "./User";
import Signin from "./Signin";

const PleaseSignIn = ({ children }) => {
  const { loading, data } = useQuery(CURRENT_USER_QUERY);
  if (loading) return <p>Loading ...</p>;
  if (!data || !data.me) return <Signin />;
  return <div>{children}</div>;
};

export default PleaseSignIn;
