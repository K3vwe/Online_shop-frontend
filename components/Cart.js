import React from "react";
import gql from "graphql-tag";
import { useMutation, useQuery } from "@apollo/react-hooks";
import CartStyles from "./styles/CartStyles";
import User from "./User";
import Supreme from "./styles/Supreme";
import CloseButton from "./styles/CloseButton";
import CartItem from "./CartItem";
import calcTotalPrice from "../lib/calcTotalPrice";
import formatMoney from "../lib/formatMoney";
import Checkout from "./Checkout";

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`;

export const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`;

const Cart = () => {
  const { data } = useQuery(LOCAL_STATE_QUERY);
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);

  return (
    <User>
      {(userData) => {
        if (!userData || !userData.me) return null;
        const { cart } = userData.me;
        return (
          <CartStyles open={data.cartOpen}>
            <header>
              <CloseButton onClick={toggleCart} title="close">
                &times;
              </CloseButton>
              <Supreme>Your Cart</Supreme>
              <p>
                You have {cart.length} item
                {cart.length <= 1 ? "" : "s"} in your cart.
              </p>
            </header>
            <ul>
              {cart.map((cartItem) => (
                <CartItem key={cartItem.id} cartItem={cartItem} />
              ))}
            </ul>
            <footer>
              <p>{formatMoney(calcTotalPrice(cart))}</p>
              <Checkout />
            </footer>
          </CartStyles>
        );
      }}
    </User>
  );
};

export default Cart;
