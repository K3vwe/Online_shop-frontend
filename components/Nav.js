import Link from "next/link";
import { useMutation } from "@apollo/react-hooks";
import { TOGGLE_CART_MUTATION } from "./Cart";
import NavStyles from "./styles/NavStyles";
import User from "./User";
import Signout from "./Signout";
import CartCount from "./CartCount";
import CartItem from "./CartItem";

const Nav = () => {
  const [toggleCart] = useMutation(TOGGLE_CART_MUTATION);
  return (
    <User>
      {(data) => (
        <NavStyles data-test="nav">
          <Link href="/items">
            <a>shop</a>
          </Link>
          {data && data.me ? (
            <>
              <Link href="/sell">
                <a>sell</a>
              </Link>
              <Link href="/orders">
                <a>orders</a>
              </Link>
              <Link href="/me">
                <a>Account</a>
              </Link>
              <Signout />
              <button onClick={() => toggleCart()}>
                My cart{" "}
                <CartCount
                  count={data.me.cart.reduce(
                    (tally, cartItem) => tally + cartItem.quantity,
                    0
                  )}
                />
              </button>
            </>
          ) : (
            <Link href="/signup">
              <a>signup</a>
            </Link>
          )}
        </NavStyles>
      )}
    </User>
  );
};

export default Nav;
