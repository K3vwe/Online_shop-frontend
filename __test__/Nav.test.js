import { mount } from "enzyme";
import wait from "waait";
import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { MockedProvider } from "@apollo/react-testing";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import { signedInMocks, notSignedinMocks } from "./PleaseSignIn.test";
import toJson from "enzyme-to-json";

const signedInWithCartItemsMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
];

describe("<Nav/>", () => {
  it("renders a minimal nav when signed out", async () => {
    const wrapper = mount(
      <MockedProvider mocks={notSignedinMocks}>
        <Nav />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    const nav = wrapper.find('[data-test="nav"]');
    expect(toJson(nav)).toMatchSnapshot();
  });

  it("renders full Nav when signed in", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInMocks}>
        <Nav />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    const nav = wrapper.find('ul[data-test="nav"]');
    expect(nav.children().length).toBe(6);
    expect(nav.text()).toContain("Sign out");
    expect(toJson(nav)).toMatchSnapshot();
  });

  it("renders the amount of items in the cart", async () => {
    const wrapper = mount(
      <MockedProvider mocks={signedInWithCartItemsMocks}>
        <Nav />
      </MockedProvider>
    );

    await wait();
    wrapper.update();
    const nav = wrapper.find('[data-test="nav"]');
    const count = nav.find("div.count");
    expect(toJson(count)).toMatchSnapshot();
  });
});
