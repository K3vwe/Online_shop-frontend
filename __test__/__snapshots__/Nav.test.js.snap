// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out 1`] = `
Array [
  <NavStyles
    data-test="nav"
  >
    <ul
      className="NavStyles-sc-11c0d2g-0 cDTdBM"
      data-test="nav"
    >
      <Link
        href="/items"
      >
        <a
          href="/items"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          shop
        </a>
      </Link>
      <Link
        href="/signup"
      >
        <a
          href="/signup"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          signup
        </a>
      </Link>
    </ul>
  </NavStyles>,
  <ul
    className="NavStyles-sc-11c0d2g-0 cDTdBM"
    data-test="nav"
  >
    <Link
      href="/items"
    >
      <a
        href="/items"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        shop
      </a>
    </Link>
    <Link
      href="/signup"
    >
      <a
        href="/signup"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        signup
      </a>
    </Link>
  </ul>,
]
`;

exports[`<Nav/> renders full Nav when signed in 1`] = `
<ul
  className="NavStyles-sc-11c0d2g-0 cDTdBM"
  data-test="nav"
>
  <Link
    href="/items"
  >
    <a
      href="/items"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      shop
    </a>
  </Link>
  <Link
    href="/sell"
  >
    <a
      href="/sell"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      sell
    </a>
  </Link>
  <Link
    href="/orders"
  >
    <a
      href="/orders"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      orders
    </a>
  </Link>
  <Link
    href="/me"
  >
    <a
      href="/me"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Account
    </a>
  </Link>
  <Signout>
    <button
      onClick={[Function]}
    >
      Sign out
    </button>
  </Signout>
  <button
    onClick={[Function]}
  >
    My cart
     
    <CartCount
      count={0}
    >
      <CartCount__AnimationStyle>
        <span
          className="CartCount__AnimationStyle-xxvp4g-0 hASYsj"
        >
          <TransitionGroup
            childFactory={[Function]}
            component="div"
          >
            <div>
              <CSSTransition
                className="count"
                classNames="count"
                in={true}
                key=".$0"
                onExited={[Function]}
                timeout={
                  Object {
                    "enter": 400,
                    "exit": 400,
                  }
                }
                unmountOnExit={true}
              >
                <Transition
                  appear={false}
                  className="count"
                  enter={true}
                  exit={true}
                  in={true}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={
                    Object {
                      "enter": 400,
                      "exit": 400,
                    }
                  }
                  unmountOnExit={true}
                >
                  <CartCount__Dot
                    className="count"
                  >
                    <div
                      className="count CartCount__Dot-xxvp4g-1 imBrrz"
                    >
                      0
                    </div>
                  </CartCount__Dot>
                </Transition>
              </CSSTransition>
            </div>
          </TransitionGroup>
        </span>
      </CartCount__AnimationStyle>
    </CartCount>
  </button>
</ul>
`;

exports[`<Nav/> renders the amount of items in the cart 1`] = `
<div
  className="count CartCount__Dot-xxvp4g-1 imBrrz"
>
  9
</div>
`;

exports[`<PleaseSignIn/> renders the sign in dialog for logged out user 1`] = `
<Signin>
  <div>
    <Form
      method="POST"
      onSubmit={[Function]}
    >
      <form
        className="Form-sc-1xszr8q-0 ekayYU"
        method="POST"
        onSubmit={[Function]}
      >
        <fieldset
          aria-busy={false}
          disabled={false}
        >
          <h2>
            Sign into your account
          </h2>
          <DisplayError
            error={Object {}}
          />
          <label
            htmlFor="email"
          >
            Email
            <input
              name="email"
              onChange={[Function]}
              placeholder="enter your email here"
              type="email"
              value=""
            />
          </label>
          <label
            htmlFor="password"
          >
            Password
            <input
              name="password"
              onChange={[Function]}
              placeholder="enter your password here"
              type="password"
              value=""
            />
          </label>
          <button
            className="button"
          >
             Sign in
          </button>
        </fieldset>
      </form>
    </Form>
  </div>
</Signin>
`;
